game/state/rules/aequipmenttype.h

```diff
@@ -118,6 +118,7 @@ class AEquipmentType : public StateObject
 
 	// Ammo, Weapons & Grenades with built-in ammo, General with charge
 	int max_ammo = 0;
+	int burst = 0;
```

Comment:
WRT the previous "default" commit, changing this to "=1" should mean that all instances of the AEquipmentType that don't explicitly specify a burst value in the xml should have it set to "1", and as the "bust" member currently doesn't exist, then no xml should contain said value, so shouldn't be an issue?

Then we can remove the "0 is actually 1" workaround in the firing code.

Or was this what you tried previously, and there's something else that's overwriting it with a '0', or otherwise modifying the value? If so, then /that/ is probably a bug.

Reply:
The problem only occurs when lazy modding (or use of editors) fails to remove the <burst> tag from a piece of equipment with it specified using a "0" instead of removing the line completely.

In these instances, if a 0 value is specified, instead of complete removal of the <burst> line. The game crashes. It needs to be told that if a 0 is explicitly given, this is in fact a 1.

For example, we have a weapon like the one below added, the modder decides they do not want burst, and instead of removing the line completely just specifies "0".

Without the code to recognise 0 as a 1, OpenApoc will CTD

```
<entry>
		<key>AEQUIPMENTTYPE_GAUSS_PISTOL_CLIP</key>
		<value>
			<type>Ammo</type>
			<id>AEQUIPMENTTYPE_GAUSS_PISTOL_CLIP</id>
			<name>Surplus Gauss Pistol Clip</name>
			<weight>1</weight>
			<held_image_pack>BATTLEUNITIMAGEPACK_item16</held_image_pack>
			<dropped_sprite>PCK:xcom3/tacdata/gameobj.pck:xcom3/tacdata/gameobj.tab:16</dropped_sprite>
			<dropped_offset>
				<x>23</x>
				<y>34</y>
			</dropped_offset>
			<dropped_shadow_sprite>PCKSHADOW:xcom3/tacdata/oshadow.pck:xcom3/tacdata/oshadow.tab:16</dropped_shadow_sprite>
			<shadow_offset>
				<x>23</x>
				<y>6</y>
			</shadow_offset>
			<equipscreen_sprite>ufodata/pequip/gausspistolclip.png</equipscreen_sprite>
			<equipscreen_size>
				<x>1</x>
				<y>1</y>
			</equipscreen_size>
			<manufacturer>ORG_SUPERDYNAMICS</manufacturer>
			<store_space>5</store_space>
			<armor>40</armor>
			<score>1</score>
			<research_dependency>
				<type>All</type>
			</research_dependency>
			<weapon_types>
				<entry>AEQUIPMENTTYPE_GAUSS_PISTOL</entry>
			</weapon_types>
			<max_ammo>20</max_ammo>
			<burst>0</burst>
			<speed>60</speed>
			<projectile_sprites>
				<entry>bulletsprites/battle/35.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/23.png</entry>
				<entry>bulletsprites/battle/24.png</entry>
				<entry>bulletsprites/battle/25.png</entry>
				<entry>bulletsprites/battle/26.png</entry>
				<entry>bulletsprites/battle/27.png</entry>
				<entry>bulletsprites/battle/28.png</entry>
				<entry>bulletsprites/battle/29.png</entry>
				<entry>bulletsprites/battle/30.png</entry>
				<entry>bulletsprites/battle/31.png</entry>
				<entry>bulletsprites/battle/32.png</entry>
				<entry>bulletsprites/battle/33.png</entry>
				<entry>bulletsprites/battle/34.png</entry>
				<entry>bulletsprites/battle/35.png</entry>
				<entry>bulletsprites/battle/36.png</entry>
				<entry>bulletsprites/battle/37.png</entry>
				<entry>bulletsprites/battle/38.png</entry>
				<entry>bulletsprites/battle/39.png</entry>
			</projectile_sprites>
			<damage>45</damage>
			<accuracy>40</accuracy>
			<fire_delay>240</fire_delay>
			<tail_size>36</tail_size>
			<range>900</range>
			<ttl>2700</ttl>
			<explosion_graphic>DOODAD_23_PLASMA</explosion_graphic>
			<fire_sfx>RAWSOUND:rawsound/tactical/weapons/gausspistol1.wav:11025</fire_sfx>
			<impact_sfx>RAWSOUND:rawsound/tactical/weapons/gausshit1.wav:11025</impact_sfx>
			<damage_type>DAMAGETYPE_ARMOR_PIERCING</damage_type>
			<trigger_type>Timed</trigger_type>
		</value>
	</entry>
```

Reply:
but currently, no mods should have a <burst> tag, as it currently doesn't do anything (and is ignored and not serialized out in the save).

So someone has to /add/ an invalid burst tag, which feels like something we can't really 'defend' against anyway - you can always break a mod by changing the data of the mod :)

Reply:
It was something i noticed when using the EWM with builds for this PR

A couple of modders are starting to take copies of the EWM and make their own changes and with that in mind, im preempting someone just changing values to 0 sooner or later

Particularly as one of the main reasons for this PR is to add OG functionality to the 52mm Cannon and allow me to make burst weapons in the battlescape (which OG does not have).

To replicate the easiest thing to do is change the Airguard 52mm Cannon Ammo from <burst>2</burst> as this PR makes it and give a <burst>0</burst> value. It also means you don't have to set up a battlescape situation to replicate, just send out the starting stormdog to attack something when you start a new game. OpenApoc will CTD when it is fired.

In short, i know to remove the line completely to avoid a CTD if we default to 1 as you wish

When things get out into the real world, i cannot trust others to know that

So as a fail-safe (and to reduce answering questions about why XXX doesn't work or causes a CTD) it's easier for the code to just effectively go "there's a 0, this is actually a 1" and not have to deal with the questions from modders later

If you wish, i can change this (after-all im still learning this stuff, so have no real grounding on the right thing to do), but i do fear where it will head without a correction of 0/null values and modders tending to just C&P stuff modifying afterwards.

Reply:
I'd very much prefer not to handle "invalid" values in code with a "We (probably) know what you actually meant" - as it can add a lot of complexity going forward. Trying to defend against every possible mistake is how you get into a mire of guessing what the user /wanted/, instead of just getting them to fix it. And by allowing "bad" values today, that just means you'll have to allow them forever, otherwise people will whine about breaking compatibility then.

But a clear error message might be useful (e.g. if (burst == 0) { LogError("Weapon WEAPON_ID has an invalid burst value of 0");})

Then there's no excuse - best to make sure things are 'correct' in the fist place, and give the mod dev as clear an idea of what's wrong, than silently seat and modify bad values.

Generally, 'data validation' should be done at data read time, not 50 callstacks deep where you'll never see it again. Maybe we should have some 'validity checks' in the serialization code?

And trying to 'correct' bad input data is the route to madness, so starting down that path would be a mistake IMHO.

Reply:
Agreed, i'll remove the argument from both agent and vehicle equipment

```
if (number_of_shots == 0)
	{
		number_of_shots = 1;
	}
```

Removed from aequipment.cpp and vequipment.cpp :)